test/unit-tests/allocator/context_manager/segment_ctx/segment_ctx_mock.h

```diff
@@ -53,4 +59,181 @@ class MockSegmentCtx : public SegmentCtx
     MOCK_METHOD(bool, UpdateStripeCount, (StripeId lsid, int logGroupId), (override));
 };
 
+class SegmentCtxMockBase : public SegmentCtx
+{
+public:
+    SegmentCtxMockBase(void)
+    {
+        addressInfo = new NiceMock<AddressInfoMock>;
+        gcContext = new NiceMock<MockGcCtx>;
+        rebuildContext = new NiceMock<MockRebuildCtx>;
+        tp = new NiceMock<MockTelemetryPublisher>;
+
+        realCtx = new SegmentCtx(tp, rebuildContext, addressInfo, gcContext, arrayId, nullptr);;
+
+        MakeMockFunctions();
+    };
+
+    virtual ~SegmentCtxMockBase(void)
+    {
+        delete realCtx;
+        delete rebuildContext;
+        delete gcContext;
+        delete addressInfo;
+        delete tp;
+    };
+
+    MOCK_METHOD(void, Init, (), (override));
+    MOCK_METHOD(void, Dispose, (), (override));
+    MOCK_METHOD(SegmentId, AllocateFreeSegment, (), (override));
+    MOCK_METHOD(uint64_t, GetNumOfFreeSegment, (), (override));
+    MOCK_METHOD(int, GetAllocatedSegmentCount, (), (override));
+    MOCK_METHOD(void, ValidateBlks, (VirtualBlks blks), (override));
+    MOCK_METHOD(bool, InvalidateBlks, (VirtualBlks blks, bool allowVictimSegRelease), (override));
+    MOCK_METHOD(void, ValidateBlocksWithGroupId, (VirtualBlks blks, int logGroupId), (override));
+    MOCK_METHOD(uint32_t, GetValidBlockCount, (SegmentId segId), (override));
+    MOCK_METHOD(int, GetOccupiedStripeCount, (SegmentId segId), (override));
+    MOCK_METHOD(bool, UpdateOccupiedStripeCount, (StripeId lsid), (override));
+    MOCK_METHOD(SegmentInfo*, GetSegmentInfos, (), (override));
+
+    inline uint32_t GetNumOfSegment(void) { return addressInfo->GetnumUserAreaSegments(); }
+    inline uint32_t GetNumOfBlksPerStripe(void) { return addressInfo->GetblksPerStripe(); }
+    inline uint32_t GetStripesPerSegment(void) { return addressInfo->GetstripesPerSegment(); }
+
+protected:
+    NiceMock<AddressInfoMock>* addressInfo;
+    NiceMock<MockGcCtx>* gcContext;
+    NiceMock<MockRebuildCtx>* rebuildContext;
+    NiceMock<MockTelemetryPublisher>* tp;
+
+private:
+    virtual void MakeMockFunctions(void)
+    {
+        MockInit();
+        MockDispose();
+        MockAllocateFreeSegment();
+        MockGetNumOfFreeSegment();
+        MockGetAllocatedSegmentCount();
+        MockValidateBlocksWithGroupId();
+        MockGetValidBlockCount();
+        MockGetOccupiedStripeCount();
+        MockUpdateOccupiedStripeCount();
+        MockGetSegmentInfos();
+    };
+
+    virtual void MockDispose(void)
+    {
+        ON_CALL(*this, Dispose).WillByDefault([this]()
+        {
+            realCtx->Dispose();
```

Comment:
결국 realCtx를 부르는데 이 mock이 어떤 용도로 사용되는지 이해가 잘 안되는데요,
설명 부탁드립니다 ㅠㅠ

Reply:
real ctx 생성하는데 들어가는 중복 코드들을 모두 없애고 target 에 대한 action들을 한 곳에 두고 이를 상속으로 변형할 수 있게해서  좀더 쉽게 UT 를 만들 수 있게 하려는 목적...

Reply:
Fixture랑 비슷한 건가요 ?.? Mock이라고 하면 real은 아니어야 할 것 같은 느낌입니다 ㅎㅎ

Reply:
일종의 spy object 를 만드는 것으로 보면 될까요??? real object 랑 똑같은 동작을 수행하는 객체를 만든 다음에, 각 test case에서 원하는 method만 on_call이나 expect로 stub을 만드는 방식이요 ㅎ 사실 자바/코틀린에서는 `Mockito.spy(ClassType)` (https://www.baeldung.com/mockito-spy) 로 spy 를 한번에 만들어 주는데 reflection 기능이 제한된 c++에서는 다른 방법이 보이지 않네요 ㅜ

Reply:
spy object 몇개를 만드는 것은 수동으로 어떻게든 할 수 있을거 같은데, integration test case를 늘리면 결국 거의 모든 class에 대해서 비슷한 코드들을 만들 수 밖에 없으려나요;

Reply:
> Fixture랑 비슷한 건가요 ?.? Mock이라고 하면 real은 아니어야 할 것 같은 느낌입니다 ㅎㅎ

Fixture랑은 다르고, Mock에서 Real object를 부르는 방식입니다. https://google.github.io/googletest/gmock_cook_book.html 여기에서 참고했어요.

"You can use the delegating-to-real technique to ensure that your mock has the same behavior as the real object while retaining the ability to validate calls. This technique is very similar to the [delegating-to-fake](https://google.github.io/googletest/gmock_cook_book.html#DelegatingToFake) technique, the difference being that we use a real object instead of a fake."

Reply:
> 일종의 spy object 를 만드는 것으로 보면 될까요??? real object 랑 똑같은 동작을 수행하는 객체를 만든 다음에, 각 test case에서 원하는 method만 on_call이나 expect로 stub을 만드는 방식이요 ㅎ 사실 자바/코틀린에서는 `Mockito.spy(ClassType)` (https://www.baeldung.com/mockito-spy) 로 spy 를 한번에 만들어 주는데 reflection 기능이 제한된 c++에서는 다른 방법이 보이지 않네요 ㅜ

네 말씀하신거랑 비슷한 느낌입니다. 원하는 action에 대해서, real object에서의 동작을 검증할 수 있도록 하는 것 인데, 
이 부분을 잘 사용하면, 여러 개의 real object (를 감싸는 mock과 함께) 들로 구성된 integration test에서 fault injection을 좀 더 수월하게 할 수 있지 않을까? 생각이 들고요, XXXXMockBase라고 이름 지어둔 클래스가 여러 개 생기기는 하는데, base class는 단순 mapping 작업이라서 어렵지는 않을 것 같습니다. 귀찮을 뿐.

Reply:
mock base class의 이름 수정을 위해서 comment 하나 살려 두겠습니다.

Reply:
file 이름을 안 고쳐서 다시 수정해서 올렸습니다. ^^ resolve 합니다